cirq-core/cirq/contrib/bayesian_network/bayesian_network_gate.py

```diff
@@ -0,0 +1,208 @@
+                # Indicates that p(q2 = 1 | q0 = 1 and q1 = 1) = 0.4
+        ]
+
+        By convention, all the probabilties are for the variable being equal to 1 and the
+        probability of being equal to zero can be inferred. In the example above, we thus have:
+        p(q2 = 0 | q0 = 1 and q1 = 0) = 1.0 - p(q2 = 1 | q0 = 1 and q1 = 0) = 0.7
+
+        Note that there is NO checking that the chain of probability creates a directed acyclic
+        graph. In particular, the order of the elements in arc_probs matters. Also, if you want to
+        specify the dependent probabilities outside of this gate, you can mark all the variables as
+        dependent in init_probs.
+
+        init_prob: A list of tuples, each representing a single variable. The first element of the
+            tuples is a string representing the name of the variable. The second element of the
+            tuples is either None for dependent variables, or a float representing a probability.
+
+        arc_probs: A list of tuples, each representing a dependence. The first element of the tuples
+            is a string representing the name of the variable. The second element of the tuples is
+            itself a tuple of n strings, representing the dependence. The third element of the
+            tuples is a list of 2**n floats, each representing the probabilities.
+
+        Raises:
+            ValueError: If the probabilities are not in [0, 1], or an incorrect number of
+            probability is specified, or if the parameter names are no passed as a tuple.
+        """
+        for _, init_prob in init_probs:
+            if init_prob is None:
+                continue
+            if init_prob < 0.0 or init_prob > 1.0:
+                raise ValueError('Initial prob should be between 0 and 1.')
+        self._init_probs = init_probs
+        for _, params, cond_probs in arc_probs:
+            if not isinstance(params, tuple):
+                raise ValueError('Conditional prob params must be a tuple.')
+            if len(cond_probs) != 2 ** len(params):
+                raise ValueError('Incorrect number of conditional probs.')
+            for cond_prob in cond_probs:
+                if cond_prob < 0.0 or cond_prob > 1.0:
+                    raise ValueError('Conditional prob should be between 0 and 1.')
+        self._arc_probs = arc_probs
+
+    def _decompose_(self, qubits: Sequence['raw_types.Qid']) -> 'cirq.OP_TREE':
+        parameter_names = [init_prob[0] for init_prob in self._init_probs]
+        qubit_map = dict(zip(parameter_names, qubits))
+
+        for param, init_prob in self._init_probs:
+            yield self._generate_got_set_for_init_prob(qubit_map[param], init_prob)
+
+        for target, params, cond_probs in self._arc_probs:
+            yield self._generate_gate_set_for_arc_prob(
+                qubit_map[target], [qubit_map[param] for param in params], cond_probs
+            )
+
+    @classmethod
```

Comment:
why use classmethods?

I think we should only use class methods if is actually used.  Otherwise I make these top level methods or methods on the class (i.e. with (self,..)).

Reply:
Removed class methods and made them top-level methods instead. Done I think.